File: sdk/php/src/DSNParser.php
<?php

declare(strict_types=1);

namespace WorkerSQL;

/**
 * DSN Parser for WorkerSQL connection strings
 */
class DSNParser
{
    /**
     * Parse a WorkerSQL DSN string
     * 
     * @param string $dsn Connection string to parse
     * @return array Parsed DSN components
     * @throws ValidationException If DSN format is invalid
     */
    public static function parse(string $dsn): array
    {
        if (empty($dsn)) {
            throw new ValidationException('INVALID_QUERY', 'DSN must be a non-empty string');
        }

        $parsed = parse_url($dsn);

        if ($parsed === false) {
            throw new ValidationException('INVALID_QUERY', 'Invalid DSN format: ' . $dsn);
        }

        // Validate protocol
        if (!isset($parsed['scheme']) || strtolower($parsed['scheme']) !== 'workersql') {
            throw new ValidationException(
                'INVALID_QUERY',
                'Invalid protocol: ' . ($parsed['scheme'] ?? 'missing') . '. Expected "workersql"'
            );
        }

        if (!isset($parsed['host'])) {
            throw new ValidationException('INVALID_QUERY', 'Host is required in DSN');
        }

        // Parse query parameters
        $params = [];
        if (isset($parsed['query'])) {
            parse_str($parsed['query'], $params);
        }

        // Extract database from path
        $database = null;
        if (isset($parsed['path']) && $parsed['path'] !== '/') {
            $database = ltrim($parsed['path'], '/');
        }

        return [
            'protocol' => $parsed['scheme'],
            'username' => isset($parsed['user']) ? urldecode($parsed['user']) : null,
            'password' => isset($parsed['pass']) ? urldecode($parsed['pass']) : null,
            'host' => $parsed['host'],
            'port' => $parsed['port'] ?? null,
            'database' => $database,
            'params' => $params,
        ];
    }

    /**
     * Get API endpoint from parsed DSN
     * 
     * @param array $parsed Parsed DSN components
     * @return string API endpoint URL
     */
    public static function getApiEndpoint(array $parsed): string
    {
        // Check if apiEndpoint is specified in params
        if (isset($parsed['params']['apiEndpoint'])) {
            return $parsed['params']['apiEndpoint'];
        }

        // Construct from host
        $protocol = ($parsed['params']['ssl'] ?? 'true') === 'false' ? 'http' : 'https';
        $port = isset($parsed['port']) ? ':' . $parsed['port'] : '';
        return "{$protocol}://{$parsed['host']}{$port}/v1";
    }
}

File: sdk/python/workersql_client/dsn_parser.py
"""
DSN Parser for WorkerSQL
Parses connection strings in the format:
workersql://[username[:password]@]host[:port][/database][?param1=value1&param2=value2]
"""

from typing import Dict, Optional
from urllib.parse import parse_qs, unquote, urlparse


class ParsedDSN:
    """Parsed DSN components"""

    def __init__(
        self,
        protocol: str,
        host: str,
        port: Optional[int] = None,
        username: Optional[str] = None,
        password: Optional[str] = None,
        database: Optional[str] = None,
        params: Optional[Dict[str, str]] = None,
    ):
        self.protocol = protocol
        self.host = host
        self.port = port
        self.username = username
        self.password = password
        self.database = database
        self.params = params or {}


class DSNParser:
    """DSN Parser for WorkerSQL connection strings"""

    @staticmethod
    def parse(dsn: str) -> ParsedDSN:
        """
        Parse a WorkerSQL DSN string
        
        Args:
            dsn: Connection string to parse
            
        Returns:
            ParsedDSN: Parsed DSN components
            
        Raises:
            ValueError: If DSN format is invalid
        """
        if not dsn or not isinstance(dsn, str):
            raise ValueError("DSN must be a non-empty string")

        try:
            parsed = urlparse(dsn)
        except Exception as e:
            raise ValueError(f"Invalid DSN format: {e}")

        # Validate protocol
        if parsed.scheme.lower() != "workersql":
            raise ValueError(
                f"Invalid protocol: {parsed.scheme}. Expected 'workersql'"
            )

        if not parsed.hostname:
            raise ValueError("Host is required in DSN")

        # Parse query parameters
        params: Dict[str, str] = {}
        if parsed.query:
            parsed_qs = parse_qs(parsed.query)
            # Convert list values to single strings
            for key, values in parsed_qs.items():
                if values:
                    params[key] = values[0]

        # Extract database from path
        database = None
        if parsed.path and parsed.path != "/":
            database = parsed.path.lstrip("/")

        return ParsedDSN(
            protocol=parsed.scheme,
            username=unquote(parsed.username) if parsed.username else None,
            password=unquote(parsed.password) if parsed.password else None,
            host=parsed.hostname,
            port=parsed.port,
            database=database,
            params=params,
        )

    @staticmethod
    def get_api_endpoint(parsed: ParsedDSN) -> str:
        """
        Extract API endpoint from DSN parameters or construct from host
        
        Args:
            parsed: Parsed DSN components
            
        Returns:
            str: API endpoint URL
        """
        # Check if apiEndpoint is specified in params
        if "apiEndpoint" in parsed.params:
            return parsed.params["apiEndpoint"]

        # Construct from host
        protocol = "http" if parsed.params.get("ssl") == "false" else "https"
        port = f":{parsed.port}" if parsed.port else ""
        return f"{protocol}://{parsed.host}{port}/v1"
